### PR TITLE
Update k8s.io/test-infra dependency and fix GCP Mason

### PR DESCRIPTION
**Pull request by ixdy**
Some recent changes to Boskos changed some of the CRDs slightly, which is now causing older Mason implementations to fail to register:
```
  error: "no kind "DRLCObjectList" is registered for version "boskos.k8s.io/v1" in scheme "pkg/runtime/scheme.go:101""   
```

Thankfully, the changes are fairly straightforward, and we can easily use a caching kubernetes client now, too. (The benefits of caching are probably fairly low, since DRLCs don't change that often.)

Relevant upstream changes for context:
https://github.com/kubernetes/test-infra/pull/16206
https://github.com/kubernetes/test-infra/pull/16249
https://github.com/kubernetes/test-infra/pull/16367

### CODE REVIEW

File: go.mod
Comment (fejta) [inline go.mod:34]: boom!

Also looks like you need to update the replace stanzas above

Comment (ixdy) [reply go.mod:34]: The issue seemed to be the github.com/Azure/go-autorest dependency. I copied the replace directive from kubernetes/test-infra, which seems to have made this work.

**Comment by ixdy** [comment]
I additionally needed to update some unrelated files, since some Prow internal data structures have changed.

**Comment by clarketm** [comment]
> I additionally needed to update some unrelated files, since some Prow internal data structures have changed.

Thanks! Additionally, can you run `make gen` then commit the changes.

**Comment by ixdy** [comment]
it looks like it's unhappy about licenses, and I have no idea how to fix this:
```
ERROR: Some modules have unrecognized licenses:
  github.com/DataDog/zstd: similar to NCSA, 63.67% confidence, path '/home/prow/go/pkg/mod/github.com/!data!dog/zstd@v1.4.1/LICENSE'
  github.com/emirpasic/gods: path '/home/prow/go/pkg/mod/github.com/emirpasic/gods@v1.12.0/LICENSE'
  github.com/kevinburke/ssh_config: path '/home/prow/go/pkg/mod/github.com/kevinburke/ssh_config@v0.0.0-20190725054713-01f96b0aa0cd/LICENSE'
  github.com/tektoncd/pipeline: path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/ghodss/yaml/LICENSE'
  github.com/tektoncd/pipeline: similar to BSD-3-Clause-Clear, 89.66% confidence, path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/gogo/protobuf/LICENSE'
  github.com/tektoncd/pipeline: path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/jmespath/go-jmespath/LICENSE'
  github.com/tektoncd/pipeline: path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/sigs.k8s.io/yaml/LICENSE'
ERROR: Some modules have no discernible license:
  github.com/otiai10/curr
  github.com/pelletier/go-buffruneio
common/Makefile.common.mk:64: recipe for target 'lint-licenses' failed
```

**Comment by clarketm** [comment]
> it looks like it's unhappy about licenses, and I have no idea how to fix this:
> 
> ```
> ERROR: Some modules have unrecognized licenses:
>   github.com/DataDog/zstd: similar to NCSA, 63.67% confidence, path '/home/prow/go/pkg/mod/github.com/!data!dog/zstd@v1.4.1/LICENSE'
>   github.com/emirpasic/gods: path '/home/prow/go/pkg/mod/github.com/emirpasic/gods@v1.12.0/LICENSE'
>   github.com/kevinburke/ssh_config: path '/home/prow/go/pkg/mod/github.com/kevinburke/ssh_config@v0.0.0-20190725054713-01f96b0aa0cd/LICENSE'
>   github.com/tektoncd/pipeline: path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/ghodss/yaml/LICENSE'
>   github.com/tektoncd/pipeline: similar to BSD-3-Clause-Clear, 89.66% confidence, path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/gogo/protobuf/LICENSE'
>   github.com/tektoncd/pipeline: path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/jmespath/go-jmespath/LICENSE'
>   github.com/tektoncd/pipeline: path '/home/prow/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/sigs.k8s.io/yaml/LICENSE'
> ERROR: Some modules have no discernible license:
>   github.com/otiai10/curr
>   github.com/pelletier/go-buffruneio
> common/Makefile.common.mk:64: recipe for target 'lint-licenses' failed
> ```
 I am not entirely certain either. It is either a bug with the license resolution logic or an issue with the dependencies. Related discussion: https://github.com/istio/istio/pull/20379. @howardjohn or @sdake any idea how to resolve?

**Comment by ixdy** [comment]
ok, I think I've fixed some of these issues by pinning to newer versions (which include the LICENSE files), though a bunch are still failing; I suspect they'll need to be whitelisted:

```
ERROR: Some modules have unrecognized licenses:
  github.com/DataDog/zstd: similar to NCSA, 63.67% confidence, path '/go/pkg/mod/github.com/!data!dog/zstd@v1.4.1/LICENSE'
  github.com/emirpasic/gods: path '/go/pkg/mod/github.com/emirpasic/gods@v1.12.0/LICENSE'
  github.com/kevinburke/ssh_config: path '/go/pkg/mod/github.com/kevinburke/ssh_config@v0.0.0-20190725054713-01f96b0aa0cd/LICENSE'
  github.com/otiai10/curr: path '/go/pkg/mod/github.com/otiai10/curr@v0.0.0-20190513014714-f5a3d24e5776/LICENSE'
  github.com/tektoncd/pipeline: path '/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/ghodss/yaml/LICENSE'
  github.com/tektoncd/pipeline: similar to BSD-3-Clause-Clear, 89.66% confidence, path '/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/gogo/protobuf/LICENSE'
  github.com/tektoncd/pipeline: path '/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/github.com/jmespath/go-jmespath/LICENSE'
  github.com/tektoncd/pipeline: path '/go/pkg/mod/github.com/tektoncd/pipeline@v0.10.1/third_party/sigs.k8s.io/yaml/LICENSE'
```

Going one-by-one:
- [github.com/DataDog/zstd](https://github.com/istio/test-infra/pull/2447/files#diff-6abb32fd9d743ffc352ee37d5968e3df) looks like a weird BSD?
- [github.com/emirpasic/gods](https://github.com/istio/test-infra/pull/2447/files#diff-6c341239e07707e8a152517557a6e842) has multiple licenses
- [github.com/kevinburke/ssh_config](https://github.com/istio/test-infra/pull/2447/files#diff-5b803b4bc1332b020150301e821fdde1) has multiple licenses
- [github.com/otiai10/curr](https://github.com/istio/test-infra/pull/2447/files#diff-ff3323f47f81443554c8988a3f5aa290) uses a DWTFYWTP license (annoyingly)
- [github.com/tektoncd/pipeline/third_party/github.com/ghodss/yaml](https://github.com/istio/test-infra/pull/2447/files#diff-f57ddd50dce7adc767130ceda9cde0fb) has multiple licenses
- [github.com/tektoncd/pipeline/third_party/github.com/gogo/protobuf](https://github.com/istio/test-infra/pull/2447/files#diff-e534dedd0efc69116195741e2bb59d93) has a slightly strange BSD
- [github.com/tektoncd/pipeline/third_party/github.com/jmespath/go-jmespath](https://github.com/istio/test-infra/pull/2447/files#diff-1d8d4422821724a221e05c152582c9fe) has an abbreviated Apache2
- [github.com/tektoncd/pipeline/third_party/sigs.k8s.io/yaml/LICENSE](https://github.com/istio/test-infra/pull/2447/files#diff-5e2600a9b62881e28834cac7cbfc0b32) has multiple licenses

**Comment by ixdy** [comment]
All of the third-party dependencies under tektoncd/pipeline are already whitelisted, just not with their third-party paths. I'm not sure whether the tool should be taught to recognize this, or if we should whitelist them explicitly.

Any guidance on what I should do about the other dependencies?

**Comment by clarketm** [comment]
> All of the third-party dependencies under tektoncd/pipeline are already whitelisted, just not with their third-party paths. I'm not sure whether the tool should be taught to recognize this, or if we should whitelist them explicitly.
> 
> Any guidance on what I should do about the other dependencies?

Whitelisting those _specific_ modules should be fine.

File: licenses/github.com/otiai10/curr/LICENSE
Comment (howardjohn) [inline licenses/github.com/otiai10/curr/LICENSE]: I am pretty sure this license is explicitly banned within google (and likely others as well)

Comment (clarketm) [reply licenses/github.com/otiai10/curr/LICENSE]: > Can we drop the dependency?

This if from `kubernetes/test-infra` https://github.com/kubernetes/test-infra/commit/f8d0138d7ad4bd1f11302d0171cbaa9d8ae42ec3. I am not sure removing it is an (easy) option. Is there an alternative besides removing dependency upstream?

**Comment by howardjohn** [comment]
I am no expert, but I don't think k8s can/should depend on it either...
Unless it's different due to CNCF

**Comment by ixdy** [comment]
I'm trying to understand where this dependency is coming from, but I'm failing.

Nothing in `k8s.io/test-infra` directly depends on it. If I try downloading the deps into `vendor/`, nothing there depends on it. `go mod why` isn't helping either.

**Comment by clarketm** [comment]
I have not traced it all the way but it is from `github.com/tektoncd/pipeline v0.10.1`

**Comment by howardjohn** [comment]
Ahh thats right... we have had this problem before. K8s license checker checks code it uses from vendor, which is pruned, whereas istio's is checking all transitive dependencies by `go mod graph`.

github.com/tektoncd/plumbing > github.com/google/go-licenses > github.com/otiai10/copy > github.com/otiai10/mint > github.com/otiai10/curr

I don't know what we do now.

**Comment by ixdy** [comment]
I opened https://github.com/otiai10/curr/issues/2. We'll see if the maintainer is willing to switch the license.

File: go.mod
Comment (clarketm) [inline go.mod]:
```suggestion
	github.com/otiai10/curr => github.com/otiai10/curr v1.0.0 // for LICENSE
```

**Review by clarketm — APPROVED** [review]

File: go.mod
Comment (clarketm) [reply go.mod]: Otherwise lgtm, once https://github.com/istio/common-files/pull/206 is merged.

**Comment by ixdy** [comment]
oh, this is going to need me to re-run `make gen`, right.

**Comment by ixdy** [comment]
ok, looks like this is hopefully set once https://github.com/istio/common-files/pull/206 gets in.

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
thanks! sorry for the license issues. the common files update should automatically get merged in here in a bit

**Comment by clarketm** [comment]
@ixdy - rebase to pull in the new `common/config/license-lint.yml`, then I think all is set.

**Comment by howardjohn** [comment]
/retest

should work without rebase

**Comment by ixdy** [comment]
ok, the last remaining question: does anything automatically build and publish `gcr.io/istio-testing/mason`? or is that a manual process?

**Comment by clarketm** [comment]
> ok, the last remaining question: does anything automatically build and publish `gcr.io/istio-testing/mason`? or is that a manual process?

It is currently manual. `make -C boskos mason-image` should do the trick. I can automate this for the next time (https://github.com/istio/test-infra/pull/2484).

**Comment by ixdy** [comment]
can you build and push when convenient? (doesn't need to happen this weekend.) I don't have access to the istio-testing project.

**Comment by clarketm** [comment]
> can you build and push when convenient? (doesn't need to happen this weekend.) I don't have access to the istio-testing project.

`gcr.io/istio-testing/mason:v20200307-1.5.0-19-gc06282b0`

**Pull request merged**
